### ISSUE TITLE
[BUG] - some plots disappearing/not displaying properly after update_data()/update_gfx() calls

### ISSUE THREAD

**Issue by pdanford** [highfestiva/finplot#234]
### Requirements (place an `x` in each of the `[ ]`)**
* [x] I realize finplot is not a web lib. (Hint: it's native!)
* [x] I've read the [snippets](https://github.com/highfestiva/finplot/wiki/Snippets) and not found what I'm looking for.
* [x] I've searched for any related issues and avoided creating a duplicate issue.
* [x] I've updated finplot (`pip install -U finplot`).
* [x] I've supplied the required data to run my code below.

### Code to reproduce
```python
import finplot as fplt
import pandas as pd
import io

# -- make a chart with 2 row with a plot in each --

data = """time,vol
          2020-08-01T00:00:00,134.18994254
          2020-08-01T00:05:00,28.68361201
          2020-08-01T00:10:00,95.9841699
          2020-08-01T00:15:00,25.52036882
          2020-08-01T00:20:00,32.99362203
          2020-08-01T00:25:00,272.87459159
          2020-08-01T00:30:00,65.2196942
          2020-08-01T00:35:00,74.78479433
          2020-08-01T00:40:00,46.8928428
          2020-08-01T00:45:00,42.36720822
          2020-08-01T00:50:00,22.21062411
          2020-08-01T00:55:00,39.78889563
          2020-08-01T01:00:00,33.73745861
          2020-08-01T01:05:00,36.34563807
          2020-08-01T01:10:00,36.1138157
          2020-08-01T01:15:00,40.76047654
          2020-08-01T01:20:00,22.53141467
          2020-08-01T01:25:00,18.07777984
          2020-08-01T01:30:00,29.34292546
          2020-08-01T01:35:00,34.43111442
          2020-08-01T01:40:00,15.24503765
          2020-08-01T01:45:00,26.6317692"""

df = pd.read_csv(io.StringIO(data))
df = df.astype({'time':'datetime64[ns, UTC]'})

ax1,ax2 = fplt.create_plot("improper plot problem", rows=2, maximize=False)

plot1 = fplt.plot(df["time vol".split()], ax=ax1, legend="test1")
plot2 = fplt.plot(df["time vol".split()], ax=ax2, legend="test2")

# ** df is updated with additional data here, but is omitted for clarity **

# -- now update the charts with updated df (à la issue #131 example) --
plot1.update_data(df["time vol".split()], gfx=False)
plot2.update_data(df["time vol".split()], gfx=False)

# --> these calls cause a blank chart on the ax2 plot
# (a zoom or scroll on the chart will cause the plot to reappear)
plot1.update_gfx()
plot2.update_gfx() 

fplt.show()
```

### Describe the bug
On a 2 row chart, updating the plot data will cause the second plot to not dispay properly.

This appears to be a bug introduced in commit 1190e8a (on 2021-07-08) where round() is replaced with _round(). Roll back to the previous commit and the bug disappears.

### Expected behavior
Both plots should remain visible after update_gfx() call. Currently, a manual scroll or zoom is needed to cause the second plot to appear after the update_data()/update_gfx() calls.

#### Screenshots
Bug version (current HEAD 2021-07-15 516db43):
<img width="793" alt="image" src="https://user-images.githubusercontent.com/5867795/127280115-18bafe9c-364c-4f93-a6aa-cccc2784bce4.png">

Roll back to previous commit (ccecee7):
<img width="797" alt="image" src="https://user-images.githubusercontent.com/5867795/127280614-c4bbdea2-3bc7-4936-8758-af9eac59a28f.png">

#### Reproducible in:

*OS*: macOS 11.5
*finplot version*: current HEAD (2021-07-15 516db43)
*pyqtgraph version*: pyqtgraph-0.12.1
*pyqt version*: PyQt5_Qt5-5.15.2

**Comment by highfestiva** [comment]
This works in d721294. Lemmeknow if you still experience bugs.

**Comment by pdanford** [comment]
Almost; it fixed the initial example above. But I was able to edit it to make the bug come back (see revised example below).

### Make sure we're rounding .5 up (1190e8a)
It looks like the specific problem is when _round(v) gets an arg of 0.5, ceil(v-0.5) transforms it to 0 instead of rounding to 1.

Assuming it doesn't break the reason for _round(), I can fix this bug by changing your _round(v) function to:

```python
def _round(v):
    """
    round towards +/- ∞ as appropriate
    """
    if v >= 0:
        v = floor(v + 0.5) # round up (to more positive)
    else:
        v = ceil(v - 0.5)  # round down (to more negative)

    return v
```

### Revised example:

```python
import finplot as fplt
import pandas as pd
import io

# -- make a chart with 2 row with a plot in each --

data = """time,low,high,open,close,vol
2020-08-01T00:00:00+00:00,11350,11400.96,11351.61,11391.66,134.18994254
2020-08-01T00:05:00+00:00,11377.9,11395,11394.91,11379.26,28.68361201
2020-08-01T00:10:00+00:00,11312.67,11379.26,11379.26,11332.36,95.9841699
2020-08-01T00:15:00+00:00,11334.62,11353.57,11335.46,11339.25,25.52036882
2020-08-01T00:20:00+00:00,11301.85,11338.06,11337.95,11303.02,32.99362203
2020-08-01T00:25:00+00:00,11239.42,11310.31,11303.01,11289.22,272.87459159
2020-08-01T00:30:00+00:00,11250,11285,11285,11262.84,65.2196942
2020-08-01T00:35:00+00:00,11237.82,11262.84,11262.84,11260,74.78479433
2020-08-01T00:40:00+00:00,11260,11288.32,11260,11287.55,46.8928428
2020-08-01T00:45:00+00:00,11269.17,11300,11288.17,11299.99,42.36720822
2020-08-01T00:50:00+00:00,11283.39,11300,11300,11285.74,22.21062411
2020-08-01T00:55:00+00:00,11256.6,11285.74,11285.74,11263.81,39.78889563
2020-08-01T01:00:00+00:00,11245.13,11263.82,11263.81,11253.16,33.73745861
2020-08-01T01:05:00+00:00,11252.42,11309.9,11252.42,11294.99,36.34563807
2020-08-01T01:10:00+00:00,11276.41,11295,11295,11285.58,36.1138157
2020-08-01T01:15:00+00:00,11285.57,11300.08,11285.57,11299.59,40.76047654
2020-08-01T01:20:00+00:00,11299.55,11308.01,11299.6,11306.79,22.53141467
2020-08-01T01:25:00+00:00,11303,11308.01,11306.13,11303,18.07777984
"""

df = pd.read_csv(io.StringIO(data))
df = df.astype({'time':'datetime64[ns, UTC]'})

ax1,ax2 = fplt.create_plot("improper plot problem", rows=2, maximize=False)

plot1 = fplt.candlestick_ochl(df["time open close high low".split()], ax=ax1)
plot2 = fplt.plot(df["time vol".split()], ax=ax2, legend="test2")

# ** df is updated with additional data here, but is omitted for clarity **

# -- now update the charts with updated df (à la issue #131 example) --
plot1.update_data(df["time vol".split()], gfx=False)
plot2.update_data(df["time vol".split()], gfx=False)

# --> these calls cause a blank chart on the ax2 plot
# (a zoom or scroll on the chart will cause the plot to reappear)
plot1.update_gfx()
plot2.update_gfx() 

fplt.show()
```

**Comment by highfestiva** [comment]
I just added your first part to fix 0.5.

**Comment by pdanford** [comment]
I'm still able to cause the bug to appear with some datasets. This bug is a bit tricky because the way rounding is done can mask it. But I think I have the real cause now after looking a bit.

### Referring to the `update_y_zoom()` implementation (current HEAD - 6f42e4f)
x0 and x1 are calculated by:
```
x0 = max(_round(x0-side_margin)+side_margin, -side_margin)
x1 = min(_round(x1-side_margin)+side_margin, datasrc.xlen+right_margin_candles-side_margin)
```

The rounding can sometimes cause x1 to be calculated such that `_,_,hi,lo,cnt = datasrc.hilo(x0, x1)` below that returns differing hi and lo which results in correct plot display, and sometimes hi and lo are equal, which later causes `rng = (hi-lo) / self.v_zoom_scale` to set rng to 0.

### The cause
The real culprit is near the top of `update_y_zoom()` in this block (line 872):
```
if x0 is None or x1 is None:
    tr = self.targetRect()
    x0 = tr.left()
    x1 = tr.right()
```
`self.targetRect()` is returning values e.g. `PyQt5.QtCore.QRectF(-0.5, -0.5, 1.0, 1.0)` and `PyQt5.QtCore.QRectF(0.0, 16.50074233545248, 1.0, 257.9508867590951)` during the initial calls to `update_y_zoom()` which is causing an x0,x1 width of one candle.

### The fix
So I can fix the bug by changing the above code block to:
```
if x0 is None or x1 is None:
    tr = self.targetRect()
    if tr.width() <= 1:
        # guard against invalid tr
        return
    else:
        x0 = tr.left()
        x1 = tr.right()
```

**Comment by highfestiva** [comment]
Thanks, put the edge case fix in 925e706.

**Comment by pdanford** [comment]
Yes, that's got it. And thank you for finplot!
